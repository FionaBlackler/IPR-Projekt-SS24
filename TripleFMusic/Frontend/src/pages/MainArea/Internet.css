.icons-body {
    display: flex;
    height:58.8rem
  }

  .icons-menu {
    flex: 1; /* Adjust the value to control the width ratio */
    /*background-color: lightblue; /* Just for visibility */
    }
    
  
 
  
  .icon {
    display: flex;
    flex-direction: column; /* Align items as a column */
    justify-content: center; /* Optional: Center vertically within the parent */
    align-items: center;
  }
  
  .buffer {
    height: 2rem;
  }
  
  .space {
    height: 2rem;
  }
  
  .draggable-window-internet {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
  }
  
  .internet-window {
    padding: 2rem;
    width: 80%;
    height: 85%;
    margin-left: 3rem;
    margin-top: 1rem;
  }
  
  .internetwindow-title {
    display: flex;
    justify-content: space-between;
    padding: 0;
    margin: 0;
  }
  
  .internet-window-content {
    display: flex;
    flex-direction: column;
  }
  
  .address-bar {
    flex: 1; /* F端llt den verf端gbaren Platz im Fenster */
    display: flex;
    align-items: center;
  }
  
  .address-input {
    flex-grow: 1; /* Das Eingabefeld f端llt den verf端gbaren Platz aus */
  }
  
  .separator-horizontal {
    border-top: 2px solid #ccc; /* Etwas dickere Linie */
    margin: 0.5rem 0;
  }
  
  .text-internet {
    display: flex;
    justify-content: center; /* Horizontal zentriert */
    align-items: center; /* Vertikal zentriert */
  }
  .internet-window-header{
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 5px 10px;
    box-shadow: none;
    border-radius: 0;
    cursor: move;
  }

  .internet-close-icon {
    display: inline-block;
    width: 16px;
    height: 16px;
    margin-left: -1px;
    margin-top: -1px;
    transform: rotateZ(45deg);
    position: relative;
    &:before,
    &:after {
      content: "";
      position: absolute;
      background: black;
    }
    &:before {
      height: 100%;
      width: 3px;
      left: 50%;
      transform: translateX(-50%);
    }
    &:after {
      height: 3px;
      width: 100%;
      left: 0px;
      top: 50%;
      transform: translateY(-50%);
    }
  }
  .close-icon{
    display: inline-block;
    width: 16px;
    height: 16px;
    margin-left: -1px;
    margin-top: -1px;
    transform: rotateZ(45deg);
    position: relative;
    &:before,
    &:after {
      content: "";
      position: absolute;
      background: black;
    }
    &:before {
      height: 100%;
      width: 3px;
      left: 50%;
      transform: translateX(-50%);
    }
    &:after {
      height: 3px;
      width: 100%;
      left: 0px;
      top: 50%;
      transform: translateY(-50%);
    }

  }
  .modal-header{
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 5px 10px;
    box-shadow: none;
    border-radius: 0;
    cursor: move;
  }
  .internet-text{
    text-align:center;
  }