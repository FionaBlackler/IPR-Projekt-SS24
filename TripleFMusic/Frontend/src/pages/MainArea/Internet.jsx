import React, { useState } from 'react';
import './Internet.css';
import Draggable from 'react-draggable';
import { ThemeProvider } from 'styled-components';
import original from 'react95/dist/themes/original';
import { Button, Window, WindowHeader, WindowContent, Toolbar, Separator, TextInput } from 'react95';

import aboutIcon from '../Images/icons/recycle2.png';
import galleryIcon from '../Images/icons/gallery4.png';
import addSongIcon from '../Images/icons/addsong2.png';
import homeIcon from '../Images/icons/computer3.png';
import internetexplorerIcon from '../Images/icons/internetexplorer.png';
import SnakeGame from './SnakeGame'; // Import the SnakeGame component

const Internet = () => {
  const [isModalOpen, setIsModalOpen] = useState(false);

  const handleClose = () => {
    // Implement your close logic here
  };

  const handleRetry = () => {
    // Handle retry action
  };

  const openModal = () => {
    setIsModalOpen(true);
  };

  const closeModal = () => {
    setIsModalOpen(false);
  };

  return (
    <div className="internet-body">
      <div className="icons-menu">
        {/* Your icon menu here */}
      </div>
      <div className="space" />
      <ThemeProvider theme={original}>
        <div className="draggable-window-internet">
          <Draggable handle=".internetwindow-title">
            <Window className="internet-window">
              <WindowHeader className="internet-window-header">
                <span>Microsoft Internet Explorer - [No Document]</span>
                <Button onClick={handleClose}>
                  <span className="internet-close-icon" />
                </Button>
              </WindowHeader>
              <Toolbar>
                <Button variant="menu" size="sm">
                  File
                </Button>
                <Button variant="menu" size="sm">
                  Edit
                </Button>
                <Button variant="menu" size="sm">
                  View
                </Button>
                <Button variant="menu" size="sm" disabled>
                  Favorites
                </Button>
                <Button variant="menu" size="sm">
                  Help
                </Button>
              </Toolbar>
              <div className="separator-horizontal" />
              <WindowContent className="internet-window-content">
                <div className="flex-container">
                  <Separator style={{ margin: "10px 0" }}>
                    <span>Adresse:</span>
                    <TextInput
                      className="textinput"
                      placeholder='type address here...'
                    />
                  </Separator>
                </div>

                <div className="internet-message-container">
                  <span className='internet-text'>Unfortunately no internet connection could be established</span>
                  <div className="internet-button-container">
                    <Button className="button-retry" onClick={handleRetry}>
                      â†» Retry
                    </Button>
                    <Button className="button-game" onClick={openModal}>
                      Let's play a game
                    </Button>
                  </div>
                </div>
              </WindowContent>
            </Window>
          </Draggable>
        </div>

        {isModalOpen && (
          <div className="modal-overlay">
            <div className="snake-overlay">
              <Draggable handle=".modal-header">
                <Window className="modal-window">
                  <WindowHeader className="modal-header">
                    <span>Snake Game</span>
                    <Button onClick={closeModal}>
                      <span className="close-icon" />
                    </Button>
                  </WindowHeader>
                  <WindowContent>
                    <SnakeGame onClose={closeModal} /> {/* Pass onClose function */}
                  </WindowContent>
                </Window>
              </Draggable>
            </div>
          </div>
        )}
      </ThemeProvider>
    </div>
  );
};

export default Internet;
