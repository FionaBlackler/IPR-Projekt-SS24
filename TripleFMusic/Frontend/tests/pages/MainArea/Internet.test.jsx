import React from 'react';
import { render, screen, fireEvent, act } from '@testing-library/react';
import Internet from '../../../src/pages/MainArea/Internet';
import '@testing-library/jest-dom';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));

describe('Internet Component', () => {
  const navigate = jest.fn();

  beforeEach(() => {
    require('react-router-dom').useNavigate.mockReturnValue(navigate);
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  const renderComponent = () => render(<Internet />);

  test('renders Internet component correctly', () => {
    renderComponent();
    expect(screen.getByText(/Microsoft Internet Explorer - \[No Document\]/i)).toBeInTheDocument();
    expect(screen.getByText(/Unfortunately no internet connection could be established./i)).toBeInTheDocument();
  });

  test('navigates when icons are clicked', () => {
    const { getByAltText } = renderComponent();

    fireEvent.click(getByAltText(/About/i));
    expect(navigate).toHaveBeenCalledWith('/welcome/about');

    fireEvent.click(getByAltText(/Home/i));
    expect(navigate).toHaveBeenCalledWith('/welcome/home');

    fireEvent.click(getByAltText(/Music Gallery/i));
    expect(navigate).toHaveBeenCalledWith('/welcome/musicgallery');

    fireEvent.click(getByAltText(/addsong/i));
    expect(navigate).toHaveBeenCalledWith('/welcome/addsong');

    fireEvent.click(getByAltText(/internetexplorer/i));
    expect(navigate).toHaveBeenCalledWith('/welcome/internet');
  });

  test('closes Internet window and navigates to home', () => {
    renderComponent();
    const closeButton = screen.getByRole('button', { name: '' });
    fireEvent.click(closeButton);
    expect(navigate).toHaveBeenCalledWith('/welcome/home');
  });

  test('opens and closes SnakeGame modal', () => {
    renderComponent();
    const openButton = screen.getByText(/Snake me up!/i);
    fireEvent.click(openButton);
    expect(screen.getByText(/Play with Snaki/i)).toBeInTheDocument();

    const closeModalButton = screen.getAllByRole('button', { name: '' }).filter(btn => btn.closest('.game-modal-window')).pop();
    fireEvent.click(closeModalButton);
    expect(screen.queryByText(/Play with Snaki/i)).not.toBeInTheDocument();
  });

  test('shows alert when retry button is clicked', () => {
    renderComponent();
    global.alert = jest.fn();
    const retryButton = screen.getByText('â†»');
    fireEvent.click(retryButton);
    expect(global.alert).toHaveBeenCalledWith('still no connection...');
  });

  test('plays SnakeGame correctly', () => {
    renderComponent();
    fireEvent.click(screen.getByText(/Snake me up!/i));

    // Capture the initial position of the snake
    const initialSegment = screen.getAllByTestId('snake-segment')[0];
    const initialLeft = parseInt(initialSegment.style.left);
    const initialTop = parseInt(initialSegment.style.top);

    // Move the snake to the right
    fireEvent.keyDown(window, { key: 'ArrowRight' });
    act(() => {
      jest.advanceTimersByTime(100);
    });

    // Capture the new position of the snake
    const movedSegmentRight = screen.getAllByTestId('snake-segment')[0];
    const movedLeft = parseInt(movedSegmentRight.style.left);
    expect(movedLeft).toBe(initialLeft + 20);

    // Move the snake down
    fireEvent.keyDown(window, { key: 'ArrowDown' });
    act(() => {
      jest.advanceTimersByTime(100);
    });

    // Capture the new position of the snake after moving down
    const movedSegmentDown = screen.getAllByTestId('snake-segment')[0];
    const movedTop = parseInt(movedSegmentDown.style.top);
    expect(movedTop).toBe(initialTop + 20);

    // Simulate game over by moving the snake out of bounds
    act(() => {
      jest.advanceTimersByTime(100 * 18);
    });
    expect(screen.getByText(/Game Over!/i)).toBeInTheDocument();

    // Restart the game
    fireEvent.click(screen.getByText(/Restart/i));
    expect(screen.queryByText(/Game Over!/i)).not.toBeInTheDocument();
  });

  
  test('increases score and repositions food in SnakeGame', async () => {
    renderComponent();
    fireEvent.click(screen.getByText(/Snake me up!/i));

    // Capture initial position of the food
    const initialFood = screen.getByTestId('food');
    const initialFoodLeft = parseInt(initialFood.style.left);
    const initialFoodTop = parseInt(initialFood.style.top);

    // Move the snake to the initial food position
    let snakeSegment = screen.getAllByTestId('snake-segment')[0];
    let snakeLeft = parseInt(snakeSegment.style.left);
    let snakeTop = parseInt(snakeSegment.style.top);

    // Move horizontally towards the food
    while (snakeLeft < initialFoodLeft) {
      fireEvent.keyDown(window, { key: 'ArrowRight' });
      act(() => {
        jest.advanceTimersByTime(100);
      });
      snakeSegment = screen.getAllByTestId('snake-segment')[0];
      snakeLeft = parseInt(snakeSegment.style.left);
    }

    while (snakeLeft > initialFoodLeft) {
      fireEvent.keyDown(window, { key: 'ArrowLeft' });
      act(() => {
        jest.advanceTimersByTime(100);
      });
      snakeSegment = screen.getAllByTestId('snake-segment')[0];
      snakeLeft = parseInt(snakeSegment.style.left);
    }

    // Move vertically towards the food
    while (snakeTop < initialFoodTop) {
      fireEvent.keyDown(window, { key: 'ArrowDown' });
      act(() => {
        jest.advanceTimersByTime(100);
      });
      snakeSegment = screen.getAllByTestId('snake-segment')[0];
      snakeTop = parseInt(snakeSegment.style.top);
    }

    while (snakeTop > initialFoodTop) {
      fireEvent.keyDown(window, { key: 'ArrowUp' });
      act(() => {
        jest.advanceTimersByTime(100);
      });
      snakeSegment = screen.getAllByTestId('snake-segment')[0];
      snakeTop = parseInt(snakeSegment.style.top);
    }

    // Allow the game to process the food eating and score update
    await act(async () => {
      jest.advanceTimersByTime(200);
    });

    // Verify food repositioning and score update
    const scoreElement = await screen.findByTestId('score');
    const newFood = screen.getByTestId('food');
    const newFoodLeft = parseInt(newFood.style.left);
    const newFoodTop = parseInt(newFood.style.top);

    expect(scoreElement.textContent).toBe('Your score: 1');
    expect(newFoodLeft).not.toBe(initialFoodLeft);
    expect(newFoodTop).not.toBe(initialFoodTop);
  });

  test('matches snapshot', () => {
    const { asFragment } = renderComponent();
    expect(asFragment()).toMatchSnapshot();
  });
});
