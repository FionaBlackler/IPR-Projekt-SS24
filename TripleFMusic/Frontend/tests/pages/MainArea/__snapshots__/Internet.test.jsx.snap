// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Internet Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="internet-body"
  >
    <div
      class="icons-menu"
    >
      <div
        style="height: 3rem;"
      />
      <div>
        <a
          class="icon"
        >
          <img
            alt="About"
            src="test-file-stub"
            style="margin-right: 4px;"
          />
          <span>
            About
          </span>
        </a>
      </div>
      <div
        class="buffer"
      />
      <div>
        <a
          class="icon"
        >
          <img
            alt="home"
            src="test-file-stub"
            style="margin-right: 4px;"
          />
          <span>
            Home
          </span>
        </a>
      </div>
      <div
        class="buffer"
      />
      <div>
        <a
          class="icon"
        >
          <img
            alt="Music Gallery"
            src="test-file-stub"
            style="margin-right: 4px;"
          />
          <span>
            Music Gallery
          </span>
        </a>
      </div>
      <div
        class="buffer"
      />
      <div>
        <a
          class="icon"
        >
          <img
            alt="addsong"
            src="test-file-stub"
            style="margin-right: 4px;"
          />
          <span>
            Upload new song
          </span>
        </a>
      </div>
      <div
        class="buffer"
      />
      <div>
        <a
          class="icon"
        >
          <img
            alt="internetexplorer"
            src="test-file-stub"
            style="margin-right: 4px;"
          />
          <span>
            Internet Explorer
          </span>
        </a>
      </div>
    </div>
    <div
      class="space"
    />
    <div
      class="draggable-window-internet react-draggable"
      style="transform: translate(100px,50px);"
    >
      <div
        class="sc-lcIQwB brMtzx internet-window"
      >
        <div
          class="sc-jxOSYQ cXImuS internet-window-header"
        >
          <span>
            Microsoft Internet Explorer - [No Document]
          </span>
          <button
            class="sc-dcJtft gGzOWQ"
            type="button"
            variant="default"
          >
            <span
              class="internet-close-icon"
            />
          </button>
        </div>
        <div
          class="sc-dCFGXG dsYMML"
        >
          <button
            class="sc-dcJtft ikfceh"
            type="button"
            variant="menu"
          >
            File
          </button>
          <button
            class="sc-dcJtft ikfceh"
            type="button"
            variant="menu"
          >
            Edit
          </button>
          <button
            class="sc-dcJtft ikfceh"
            type="button"
            variant="menu"
          >
            View
          </button>
          <button
            class="sc-dcJtft eBZzQn"
            disabled=""
            type="button"
            variant="menu"
          >
            Favorites
          </button>
          <button
            class="sc-dcJtft ikfceh"
            type="button"
            variant="menu"
          >
            Help
          </button>
        </div>
        <div
          class="separator-horizontal"
        />
        <div
          class="sc-fhzEvr heDfsD internet-container"
        >
          <div
            class="address-input-container"
          >
            <span>
              Address:
            </span>
            <div
              class="sc-dhKdPU sc-feUYzb gySDZy bKpwcv textinput"
              data-testid="variant-default"
            >
              <input
                class="sc-hzhKNl hFvftg"
                placeholder="Type address here..."
                type="text"
                variant="default"
              />
            </div>
            <button
              class="sc-dcJtft gGzOWQ button-retry"
              type="button"
              variant="default"
            >
              â†»
            </button>
          </div>
          <div
            class="sc-cPiJYC eIKBiO"
          />
          <div
            class="internet-message-container"
          >
            <span
              class="internet-text"
            >
              Unfortunately no internet connection could be established.
            </span>
            <div
              class="internet-button-container"
            >
              <button
                class="sc-dcJtft gGzOWQ button-game"
                type="button"
                variant="default"
              >
                Snake me up!
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
